### PR TITLE
fix matrix_io_delay() timing in quantum/matrix.c

### PR DESCRIPTION
**Pull request by mtei**
## Description

~~The timing of the call to matrix_io_delay() has been changed to the appropriate time.  With this change, we can reduce the number of times we call matrix_io_delay() by one.~~

**Separated `matrix_io_delay()` into the following two functions so that you can set the appropriate delay value for each.**

* `matrix_output_select_delay()` -  after `select_row()/select_col()`
  The delay is a small number of clocks specific to the MCU.  
  The default implementation is as follows:
  ```c
  __attribute__((weak)) void matrix_output_select_delay(void) { waitInputPinDelay(); }
  ```
   See below for more information on `waitInputPinDelay()`.

* `matrix_output_unselect_delay()` - after `unselect_row()/unselect_col()`
  The delay is in the range of a few microseconds, depending on the capacitance and resistance of the entire keyboard circuitry.  
  The default implementation is as follows:
  ```c
  __attribute__((weak)) void matrix_io_delay(void) { wait_us(MATRIX_IO_DELAY); }
  __attribute__((weak)) void matrix_output_unselect_delay(void) { matrix_io_delay(); }
  ```

**Added `waitInputPinDelay()` into `quantum/quantum.h`.**

On AVR's GPIO and ARM's GPIO, the input signal changes need some clock time to be read into the input pins.

The `waitInputPinDelay()` will wait the necessary time. The wait time is set to `GPIO_INPUT_PIN_DELAY` in units of the clock.

If `GPIO_INPUT_PIN_DELAY` is not set, the following values are used.

* AVR
  The datasheets for ATmega32u4/16u4, ATmega32u2/16u2, ATmega328p, AT90usb646/1286, etc. say that a delay of one clock is required after a change in the input signal.  Therefore, the default value of GPIO_INPUT_PIN_DELAY can be set to 1, but we'll set it to 2 to allow for some leeway.

* ARM-based MCUs
  For GPIOs on ARM-based MCUs, the input pins are sampled by the clock of the bus to which the GPIO is connected.
  The connected buses differ depending on the various series of MCUs.
  Also, since the CPU instruction execution clock and GPIO bus clock can vary depending on the MCU GPIO bus configuration and MCU internal register settings, the optimal delay value cannot be determined. Therefore, GPIO_INPUT_PIN_DELAY defaults to a rather large value of 0.25 microseconds.

<details>
<summary><strong>Current matrix.c timing (click)</strong></summary>

#### over view
old_overview

#### detail view
old_detail

</details>
<details>
<summary><strong>New matrix.c timing (click)</strong></summary>

#### over view
new_overview

#### detail view
new_detail
</details>

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mtei** [comment]
> https://github.com/qmk/qmk_firmware/blob/d4dc2a577fbf6de948336288c6ca12359b48e50b/quantum/matrix.c#L96
> 
> Given the above comment, i cant see why the matrix delay was moved from directly after `select_row`?

That comment is wrong. I think a lot of people have the same misconceptions as the comment.

> Also, was there a reason this was not applied to `quantum/split_common/matrix.c`?

Of course, I was going to change `quantum/split_common/matrix.c` as well.

File: quantum/matrix.c [hunk 1]
```diff
@@ -108,6 +107,9 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
 
     // Unselect row
     unselect_row(current_row);
+    if( current_row + 1 < MATRIX_ROWS ) {
+        matrix_io_delay(); // wait for row signal to HIGH
```
Comment (fauxpark) [inline quantum/matrix.c:111]:
```suggestion
    if (current_row + 1 < MATRIX_ROWS) {
        matrix_io_delay();  // wait for row signal to go HIGH
```

File: quantum/matrix.c [hunk 2]
```diff
@@ -170,6 +171,9 @@ static bool read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col)
 
     // Unselect col
     unselect_col(current_col);
+    if( current_col + 1 < MATRIX_COLS ) {
+        matrix_io_delay(); // wait for col signal to HIGH
```
Comment (fauxpark) [inline quantum/matrix.c:175]:
```suggestion
    if (current_col + 1 < MATRIX_COLS) {
        matrix_io_delay();  // wait for col signal to go HIGH
```

File: quantum/split_common/matrix.c [hunk 3]
```diff
@@ -121,6 +120,9 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
 
     // Unselect row
     unselect_row(current_row);
+    if( current_row + 1 < MATRIX_ROWS ) {
+        matrix_io_delay(); // wait for row signal to HIGH
```
Comment (fauxpark) [inline quantum/split_common/matrix.c:124]:
```suggestion
    if (current_row + 1 < MATRIX_ROWS) {
        matrix_io_delay();  // wait for row signal to go HIGH
```

File: quantum/split_common/matrix.c [hunk 4]
```diff
@@ -183,6 +184,9 @@ static bool read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col)
 
     // Unselect col
     unselect_col(current_col);
+    if( current_col + 1 < MATRIX_COLS ) {
+        matrix_io_delay(); // wait for col signal to HIGH
```
Comment (fauxpark) [inline quantum/split_common/matrix.c:188]:
```suggestion
    if (current_col + 1 < MATRIX_COLS) {
        matrix_io_delay();  // wait for col signal to go HIGH
```

**Comment by mtei** [comment]
## FYI:

### after `select_row()`
1 cpu clock time is needed between `select_row()` and `readPin()` .
See the following figure in the ATmega32u4 datasheet.

> Figure 10-4. Synchronization when Reading a Software Assigned Pin Value

### after `unselect_row()`

You will have to wait for the row and column signals go up to HIGH before the next  `select_row()`. 
HIGH is 1.9V(=Vcc * 0.2 + 0.9) according to the ATmega32u4 datasheet.

**Comment by zvecr** [comment]
QMK runs on more than a 32u4, we would have to certain on the observed behaviour is consistent on all supported chips. Given where the delay is right now, that seems "safer" while not optimal. Not a massive fan of the risk vs reward here.

**Comment by mtei** [comment]
Look at the picture above.

There is an unstable time for the signal just after unselect_row(). In the current code, matrix_io_delay() doesn't cover that time.

This PR is intended to remedy that.

**Comment by zvecr** [comment]
> There is an unstable time for the signal just after unselect_row(). In the current code, matrix_io_delay() doesn't cover that time.

In its current form, the delay before read will also handle the last `unselect_row`

**Comment by mtei** [comment]
> QMK runs on more than a 32u4, we would have to certain on the observed behaviour is consistent on all supported chips. 

The waveform after unselect_row() above is the result of a routine manipulation of the GPIO.  That's the basic behavior of digital circuits. All AVRs and ARMs behave no differently.

**Comment by zvecr** [comment]
> If you don't believe my explanation, ask someone who knows about hardware.

Its not my burden to provide defence and justification of the change, its the submitter. I'm asking for proof of the statement `All AVRs and ARMs behave no differently`.

If we go with your previous observation of 1 clock cycle is required after `select_row`, what if a longer delay is required on some supported chips? The only way we know for certain is through testing or relying on data sheets. All we have are words and some 32u4 scope catpures, which is not enough proof to justify the change is value for all scenarios.

**Comment by mtei** [comment]
> Its not my burden to provide defence and justification of the change, its the submitter. I'm asking for proof of the statement `All AVRs and ARMs behave no differently`.

That’s right. Okay, I will try again.

**Comment by tzarc** [comment]
> 
> If we go with your previous observation of 1 clock cycle is required after `select_row`, what if a longer delay is required on some supported chips? The only way we know for certain is through testing or relying on data sheets. All we have are words and some 32u4 scope catpures, which is not enough proof to justify the change is valid for all scenarios.

Specifically to this point, I know that at least three `nop` instructions are required on at least some models of STM32 -- there is no observable change in GPIO state (using an oscilloscope) when pulsing high/low/high, or low/high/low.

Unsure if it's anything to do with instruction pipelining or the like, but single instruction GPIO is definitely not the case for everything.

**Comment by mtei** [comment]
I downloaded one of the STM32 series datasheets and am browsing through it.

The GPIO seems to be connected to the APB2 bus, which is slower than the CPU. It's running on the clock. Perhaps that's what's involved.

**Comment by tzarc** [comment]
Just a note, been running this implementation for several weeks without any issues.

**Comment by drashna** [comment]
Should probably retarget to `develop` and rebase.

**Comment by mtei** [comment]
Due to tzarc's comment above, I'm investigating the need for a very short delay after 'select_row()'.

**Comment by drashna** [comment]
> Due to tzarc's comment above, I'm investigating the need for a very short delay after 'select_row()'.

Would it be worth to have this configurable with a define? and a different function name (eg `matrix_io_delay_row` or some such)?

**Comment by mtei** [comment]
> Would it be worth to have this configurable with a define? and a different function name (eg `matrix_io_delay_row` or some such)?

We may not need this often, but it's a good idea to have it.
* select_delay() - after select_row() / select_col()
* unselect_delay() - after unselect_row() / unselect_col()

**Comment by tzarc** [comment]
Bad stalebot.

**Review by drashna — APPROVED** [review]

**Comment by mtei** [comment]
@zvecr @tzarc 

I have not yet determined the delay value required for GPIO of ARM-based MCUs based on the datasheet. Looking at the data sheet of the STM32 series, it seems that the bus configurations of F0xx, F1xx, F3xx, and F4xx are different, and the bus clock also changes depending on the setting of `mcuconf.h`.

Experiments have confirmed that the delay value in Proton-C is 2 clocks. However, this cannot be considered a general value. (https://user-images.githubusercontent.com/2170248/99894767-d5167700-2cc9-11eb-806e-f8fb05b5558e.png)

Therefore, this PullRequest has stopped determining the optimal value. Instead, I decided to set a delay value with sufficient margin.

**Review by tzarc — APPROVED** [review]
I can see that my review was re-requested -- been using this on the Djinn for several months (including newer iterations), and the code looks and works perfectly fine as far as I can see.

**Review by drashna — APPROVED** [review]

**Comment by nomis** [comment]
Why is there no delay after the last unselect?

When the next matrix_scan runs it could select the first row/column again before 30µs has passed.

**Comment by mtei** [comment]
The default of 30us for MATRIX_IO_DELAY is probably too long.

In this figure, 5us seems to be enough.

old_detail_2

Later, when I have more time, I will measure the AVR case and the Proton-C(ARM) case.

**Comment by nomis** [comment]
Ok, but there should still be a delay after the last unselect?

Otherwise there's an assumption being made that it will take at least that amount of time for other processing before selecting again.

**Comment by mtei** [comment]
In the above picture (AVR), it looks like it takes about 40us from the end of matrix_scan() until the next matrix_scan() is called. In the case of ARM, assuming a 4x clock, it would be around 10us.

If the scan interval is shorter and insufficient, you can increase GPIO_INPUT_PIN_DELAY to extend the matrix_output_select_delay() time. You can increase the GPIO_INPUT_PIN_DELAY in units of cpu clocks, so you can adjust it precisely.

**Comment by nomis** [comment]
> In the above picture (AVR), it looks like it takes about 40us from the end of matrix_scan() until the next matrix_scan() is called.

Yes but that's in one specific case. If someone runs QMK on something much faster or decides to call matrix_scan() in a loop until the matrix changes it could fail to meet the minimum time.

In reality I can remove the unselect delay entirely and the keyboard still works ok because it's just not fast enough to be a problem.

**Comment by mtei** [comment]
The following functions in `quantum/matrix_common.c` can be overridden, so do so if necessary.
```c
__attribute__((weak)) void matrix_output_select_delay(void) { waitInputPinDelay(); }
__attribute__((weak)) void matrix_output_unselect_delay(void) { matrix_io_delay(); }
__attribute__((weak)) void matrix_scan_kb(void) { matrix_scan_user(); }
__attribute__((weak)) void matrix_scan_user(void) {}
```

**Comment by nomis** [comment]
Yes I know I can override those functions, I'm stating that the default implementation shouldn't skip the last unselect delay.

**Pull request merged**